Feature/163650272 add controls prop t sne plot

haideriqbal opened the pull request:
- Tsne Plot widget uses TsnePlotView component to render tsne plots. In Tsne plot widget we don't need to show the settings controls above the scatter plots. In order to achieve it, a new boolean prop showControls is added to TsnePlotView component to show/hide the controls depending on the flag value.

- Test cases have been added to ClusterTSnePlot and GeneExpressionTSnePlot to check if the controls hide when flag is set to false.

- Default value for prop showControls is set to true. So in principle, this change won't affect the anything where this component is already being used.

haideriqbal committed:
- Add show controls prop to show/hide plot setting dropdowns
- Add test case to check if plot setting appear if flag is false

haideriqbal committed:
3.5.0-beta

lingyun1010 commented on __test__/GeneExpressionTSnePlot.test.js:
```diff
@@ -154,4 +155,14 @@ describe(`GeneExpressionTSnePlot`, () => {
+    const wrapper = mount(<GeneExpressionTSnePlot height={600} expressionGradientColours={gradientColourRanges()} atlasUrl={``} suggesterEndpoint={``} onSelectGeneId={onSelectGeneId} loading={true} plotData={plotData(randomSeries)} highlightClusters={[]} speciesName={``} showControls={false}/>)
+
+
+    expect(wrapper.find(AtlasAutocomplete).length).toBe(0)
```
Please add a test for the existence of `AtlasAutocomplete` with default `showControls` props. Remove the extra spaces.

haideriqbal replied:
Good point. Added new test case for this

lingyun1010 commented on __test__/ClusterTSnePlot.test.js:
```diff
@@ -113,6 +113,18 @@ describe(`ClusterTSnePlot`, () => {
+      series: []
+    }
+
+    const wrapper = mount(<ClusterTSnePlot height={500} ks={[]} metadata={[]} selectedColourBy={`0`} onChangeColourBy={onChangeColourBy} perplexities={[]} selectedPerplexity={0} onChangePerplexity={onChangePerplexity} loading={true} plotData={plotData} showControls={false}/>)
```
I see the props are repeated in each test, which is not from this branch, but could you please refactor a bit with a const `props` which contains all the shared ones?

haideriqbal replied:
Refactor the code to share the common props.

lingyun1010 replied:
Actually what I mean is the props of `height`, `ks`, `metadata` etc shared by all wrappers. You could assign them as a variable and pass it to `ClusterTSnePlot` to reduce the repeat of codes. Same comment for `GeneExpressionTSnePlot.test`. Please remember to add `showControls` props explicitly in the test.

lingyun1010 replied:
Right. Could you please try to put all common props into one variable like
```
const props = {height: 500, ks:...., plotData:..., metadata:...}
```
and then pass the `props` into each tested component and add any additional props if needed?

haideriqbal replied:
Implemented as suggested.

lingyun1010 commented on src/ClusterTSnePlot.js:
```diff
@@ -209,4 +216,8 @@ ClusterTSnePlot.propTypes = {
 }
 
+ClusterTSnePlot.defaultProps = {
+  showControls: true
```
As you always pass the `showControls` from `TSnePlotView` component, you just need to clarify once in `TSnePlotView`, but do not need to specify it in `ClusterTSnePlot.defaultProps`. Same comment for `GeneExpressionTSnePlot`.

haideriqbal replied:
Default prop was added in ClusterTSnePlot and GeneExpressionTSnePlot for the test cases respectively. Otherwise I have to specify showControl explicitly in each of them which would lead to more code change as compared to introducing defaultProp.

As in the test cases, we are testing ClusterTSnePlot and GeneExpressionTSnePlot and not TSnePlotView so can't get default value from there.

lingyun1010 reviewed (changes requested):
Nice work but some tiny things need to be changed. Please see the comments in detail.

haideriqbal committed:
- Add test case to check if atlas autocomplete exist with default flag value

haideriqbal committed:
- Refactor code to share same props rather than specifying them separately in each test case

haideriqbal committed:
- Refactor code to share same props for test cases
- Show only colour by control in ClusterTSnePlot
- Remove default showControl prop from ClusterTSnePlot and GeneExpressionTSnePlot
- Add Css to conditionally align GeneExpressionTSnePlot with ClusterTSnePlot
- Re-render plots if metadata is changed by color by dropdown

lingyun1010 commented on src/ClusterTSnePlot.js:
```diff
@@ -150,39 +150,47 @@ const ClusterTSnePlot = (props) => {
-      errorMessage={errorMessage}
-    />
-  ]
+  return (
```
Why you add a pair of brackets for `return`, as it will work when we return an array without the brackets as the original codes.

haideriqbal replied:
Might have missed removing them after testing. Removed now.

lingyun1010 commented on src/GeneExpressionTSnePlot.js:
```diff
@@ -174,33 +173,39 @@ const GeneExpressionScatterPlot = (props) => {
 
 GeneExpressionScatterPlot.propTypes = {
   height: PropTypes.number.isRequired,
+  showControls: PropTypes.bool,
```
The `showControls` props in `GeneExpressionTSnePlot` and `ClusterTSnePlot` should be set as `isRequired` in this case.

haideriqbal replied:
Good point. Made them mandatory

lingyun1010 commented on src/TSnePlotView.js:
```diff
@@ -84,6 +84,8 @@ class TSnePlotView extends React.Component {
       this._fetchAndSetStateCellClusters(this.props)
     } else if (previousProps.geneId !== this.props.geneId) {
       this._fetchAndSetStateGeneId(this.props)
+    } else if (previousProps.selectedColourBy !== this.props.selectedColourBy) {
```
This condition could be combined into 
```
else if (previousProps.selectedColourByCategory !== this.props.selectedColourBy &&
previousProps.selectedColourBy !== this.props.selectedColourBy)
```
where I believe its a bug here, because `previousProps.selectedColourByCategory !== this.props.selectedColourBy ` will always be true @alfonsomunozpomer @monicajianu .

Hence I think the combined condition should be 
```
else if (previousProps.selectedColourByCategory !== this.props.selectedColourByCategory ||
previousProps.selectedColourBy !== this.props.selectedColourBy)
```

haideriqbal replied:
I agree seems like bug. Have modified the conditional statement accordingly.

lingyun1010 reviewed (commented):
Please check the comments below.

haideriqbal committed:
- Refactor code to share same props for test cases
- Make showControl required in ClusterTsnePlot and GeneExpressionTsnePlot
- Refactor condition logic for componentDidMount in TsnePlotView.js

lingyun1010 reviewed (approved):
All good. Please merge.

haideriqbal committed:
Merge branch 'master' into feature/163650272-add-controls-prop-t-sne-plot
